File: lib/Ruby/numbers.rb
class Integer
  def self.primes(options)
    @@generator ||= PrimeGenerator.new
    @@generator.upto(options[:less_than])
  end
  
  def factors()
    if self == 1
      [1]
    else
      low_factors = (1..Math.sqrt(self).round).to_a.select { |n| self % n == 0 }
      high_factors = low_factors.map { |n| self / n }.reverse
      low_factors + high_factors
    end
  end
  
  def triangle_number
    (1..self).to_a.inject(0) { |sum, n| sum + n }
  end
  
  def factorial
    acc = 1
    (1..self).each do |i|
      acc *= i
    end
    acc
  end
end

class PrimeGenerator
  def upto(n)
    primes = []
    (2..n).each do |i|
      if primes.all? { |p| i % p != 0 }
        primes << i
      end
    end
    primes
  end
end

File: lib/Haskell/Numbers.hs
module Numbers (factorial, 
                isEven,
                primes, primeFactors, factors, 
                triangleNumber) where

isEven n = n `mod` 2 == 0

primes :: [Integer]
primes = sieve [2..]
  where sieve (p:xs) = p : sieve [x | x <- xs, x `mod` p /= 0]

primeFactors 1 = []
primeFactors n = go n primes
   where
   go n ps@(p:pt)
      | p*p > n        = [n]
      | n `rem` p == 0 = p : go (n `quot` p) ps
      | otherwise      = go n pt

factors n = filter (\i -> n `mod` i == 0) [1..n]

factorial n = product [1..n]
  
triangleNumber n = sum [1..n]
